mlir-sycl/lib/Dialect/SYCL/IR/SYCLHostOps.cpp

```diff
@@ -10,3 +10,22 @@ LogicalResult SYCLHostConstructorOp::verify() {
            << type;
   return success();
 }
+
+void SYCLHostConstructorOp::getEffects(
+    SmallVectorImpl<SideEffects::EffectInstance<MemoryEffects::Effect>>
+        &effects) {
+  // NOTE: This definition assumes only the first (`this`) argument is written
+  // to and the remaining arguments, if they are memory arguments, are both read
+  // and written. The latter is a conservative assumption about the behavior of
+  // the constructors and could be changed to enable better analysis.
+  auto *defaultResource = SideEffects::DefaultResource::get();
+  // The `this` argument will always be written to
+  effects.emplace_back(MemoryEffects::Write::get(), getDst(), defaultResource);
+  // The rest of the arguments will be scalar or read and write.
+  for (auto value : getArgs()) {
+    if (isa<MemRefType, LLVM::LLVMPointerType>(value.getType())) {
+      effects.emplace_back(MemoryEffects::Read::get(), value, defaultResource);
+      effects.emplace_back(MemoryEffects::Write::get(), value, defaultResource);
```

Comment:
This is more conservative that it needs to be, the remaining pointer arguments are not written to.

Reply:
I'm unsure whether we can guarantee that, do you think so? 

For example, the constructor of `sycl::accessor` takes a non-const `sycl:buffer &`, so technically, the constructor could modify the buffer (even though I think it's unlikely).

Reply:
Technically if could, however that argument should have been declared `const &` so I think in practice it does not. 
Placing the write memory effect on pointer argument that will not be modified is a pretty heavy barrier to optimizations. 
Even if we end up not optimizing the host code for the time being I am worried it will limit analysis and make us miss opportunities to change the device code in the future.

Reply:
> Placing the write memory effect on pointer argument that will not be modified is a pretty heavy barrier to optimizations.

Yes, that was also my concern, that the write effect would hinder analysis, e.g., based on the reaching definition analysis that you added. I've removed the write effect in the latest update.